### ISSUE TITLE
angular components testing? Karma adapter?

### ISSUE THREAD

**Issue by rlasjunies** [alsatian-test/alsatian#254]
Hi @jamesrichford 

 I am interested to use angular2 and their UI component concept. In term of UT jasmine is good but it "hurt my eyes" compare to alsatian.

 I did not found alsatian karma adapter, do you know if someone (you?) worked on it?

 What is your plan for future to be linked with a test runner like karma? Do you plan to promote one or another? Do you have a preference?

 Thank

Richard

**Comment by jamesrichford** [comment]
Hey @rlasjunies,

This is one of the things we will be working on soon see #149 :) I've had a brief look at creating a karma plugin and doesn't look like too much trouble to get a basic version working.

In terms of strategy - I think that if we're talking about unit tests, these should only be run on node as the point of unit tests is to prove that the functionality you've written works and with modern UI frameworks there should be no need for a browser for testing this functionality for the most part and the odd situations where you need to use the document or window you can fake it so a runner doesn't seem necessary. Of course there are situations where you'll want to write integration tests where something like Karma will be required, but for the most part I prefer to write end to end tests to cover this sort of behavior using something like Selenium.

In terms of usage of Karma would you see this as essential or nice to have so we can better prioritise working on a plugin and would an initial plugin with minimal features be preferable to a fully fledged finished plugin?

Thanks,

James

**Comment by rlasjunies** [comment]
hi @jamesrichford,

about "In terms of usage of Karma would you see this as essential or nice to have so we can better prioritise working on a plugin and would an initial plugin with minimal features be preferable to a fully fledged finished plugin?"
=> a minimal feature is good enough at my stage

What was looking for is about examples on how to run tests on Angular2 components. I think we are aligned on the concept unit test with karma adapter

If you would like me to try few things in parallel let me know.

Sincerely,

**Comment by jamesrichford** [comment]
I would say test the class as you would any other class :)

In terms of testing as to whether bindings, event handling etc. work do those in end to end tests using Selenium or similar. The reason behind this is to test whether or not these aspects of a component are working or not means you are testing whether Angular, the browser, the template and your code is all working together in harmony. This either means a lot of faking, which doesn't really seem like it's telling you a whole lot OR a lot of setup to ensure all the test is reflective of what would actually be happening. After all these considerations and on top of this that end to end tests are essential to ensure the User Experience is correct and that everything ties up together it seems to make more sense to say well let's leave those integration tests out and instead write more End to End Tests.

Does this make sense at all? I'm writing a series of articles on testing and one of them will cover this but hopefully this brief synopsis helps. If not please let me know and I'll elaborate :D I'm hoping to get some examples of what I'm talking about done soon so hopefully with those it'll be clearer.

**Comment by jamesrichford** [comment]
To keep you updated @rlasjunies, I've earmarked 1.1.0 to include improvements for integration and end to end testing. 

This is to include;
* examples of use with Selenium
* examples of use with Karma
* asynchronous Setup and Teardown functions
* examples of using Alsatian holistically and strategies to implement different aspects of testing

**Comment by jamesrichford** [comment]
@rlasjunies after a bit of investigation there are a few considerations to getting Alsatian working with Karma so I'm moving this to it's own separate focused milestone (1.2.0). 1.1.0 will be focusing on end to end testing with Alsatian so will have examples of use with Selenium and also a more holistic strategy of testing with Alsatian :)

**Comment by rlasjunies** [comment]
I do not know karma but I can spend few hours to learn it and then, propose you some options. Do you accept help?

**Comment by jamesrichford** [comment]
Yeah of course, I've done a bit of investigation so far. I'll assign you to the issue I've got for the investigation and put my findings so far. :)

**Comment by jamesadarich** [comment]
Running in browser is not on the roadmap. Tools like `jsdom` better support this with the ethos of running these tests in node as unit or integration tests. Whereas end to end tests are better handled by tools like `puppeteer` or `selenium`. In terms of angular happy to help out with any integrations or adapters that may be useful but closing for now. Feel free to open a more specific angular issue if necessary.
